File: ir/pkg/spec.go
package ir

import (
	"encoding/json"
	"fmt"
)

// PipelineSpec represents the complete pipeline specification.
// It is immutable and versioned for audit and replay purposes.
type PipelineSpec struct {
	APIVersion string           `json:"apiVersion"`
	Kind       string           `json:"kind"`
	Metadata   PipelineMetadata `json:"metadata"`
	Tasks      map[string]*Task `json:"tasks"`
	Edges      []Edge           `json:"edges"`
}

// PipelineMetadata contains pipeline metadata.
type PipelineMetadata struct {
	Name        string            `json:"name"`
	Version     string            `json:"version,omitempty"`
	Namespace   string            `json:"namespace,omitempty"`
	Owner       string            `json:"owner,omitempty"`
	Description string            `json:"description,omitempty"`
	Tags        map[string]string `json:"tags,omitempty"`
}

// TaskType represents the type of task in the pipeline.
type TaskType string

const (
	TaskTypeSource      TaskType = "Source"
	TaskTypeTransform   TaskType = "Transform"
	TaskTypeSink        TaskType = "Sink"
	TaskTypeConditional TaskType = "Conditional"
	TaskTypeRetry       TaskType = "Retry"
	TaskTypeSchedule    TaskType = "Schedule"
)

// Task represents a node in the pipeline DAG.
type Task struct {
	Type           TaskType               `json:"type"`
	Handler        Handler                `json:"handler"`
	Description    string                 `json:"description,omitempty"`
	Inputs         map[string]Schema      `json:"inputs,omitempty"`
	Outputs        map[string]Schema      `json:"outputs,omitempty"`
	ExecutorConfig map[string]interface{} `json:"executorConfig,omitempty"`
	CostEstimate   *CostEstimate          `json:"costEstimate,omitempty"`
	Retry          *RetryPolicy           `json:"retry,omitempty"`
	Timeout        string                 `json:"timeout,omitempty"`
	Metadata       map[string]string      `json:"metadata,omitempty"`
}

// Handler specifies how a task is executed.
type Handler struct {
	Type   string            `json:"type"` // "python", "sql", "spark", "docker", "http"
	Source string            `json:"source"`
	Env    map[string]string `json:"env,omitempty"`
}

// Schema represents a JSON Schema for task ports (inputs/outputs).
// It can be any valid JSON Schema.
type Schema map[string]interface{}

// Edge represents a data flow connection between task outputs and inputs.
type Edge struct {
	From TaskPort `json:"from"`
	To   TaskPort `json:"to"`
}

// TaskPort represents a specific port on a task (input or output).
type TaskPort struct {
	Task string `json:"task"`
	Port string `json:"port"`
}

// RetryPolicy defines retry behavior.
type RetryPolicy struct {
	MaxAttempts         int     `json:"maxAttempts"`
	Backoff             string  `json:"backoff"` // "linear", "exponential", "fixed"
	BackoffMultiplier   float64 `json:"backoffMultiplier,omitempty"`
	InitialDelaySeconds int     `json:"initialDelaySeconds,omitempty"`
}

// CostEstimate represents the estimated cost of running a task.
type CostEstimate struct {
	Compute *CostDimension `json:"compute,omitempty"`
	Storage *CostDimension `json:"storage,omitempty"`
	Network *CostDimension `json:"network,omitempty"`
}

// CostDimension represents a single cost dimension.
type CostDimension struct {
	Unit     string  `json:"unit"`
	Quantity float64 `json:"quantity"`
}

// NewPipelineSpec creates a new pipeline specification.
func NewPipelineSpec(name string) *PipelineSpec {
	return &PipelineSpec{
		APIVersion: "flowforge.io/v1",
		Kind:       "Pipeline",
		Metadata: PipelineMetadata{
			Name: name,
		},
		Tasks: make(map[string]*Task),
		Edges: []Edge{},
	}
}

// Validate checks if the PipelineSpec is valid.
// It does NOT perform executor-specific validation.
func (ps *PipelineSpec) Validate() error {
	if ps.APIVersion != "flowforge.io/v1" {
		return fmt.Errorf("invalid apiVersion: %s", ps.APIVersion)
	}
	if ps.Kind != "Pipeline" {
		return fmt.Errorf("invalid kind: %s", ps.Kind)
	}
	if ps.Metadata.Name == "" {
		return fmt.Errorf("metadata.name is required")
	}
	if len(ps.Tasks) == 0 {
		return fmt.Errorf("at least one task is required")
	}

	// Validate each task exists and edges reference valid tasks/ports
	for edgeIdx, edge := range ps.Edges {
		fromTask, ok := ps.Tasks[edge.From.Task]
		if !ok {
			return fmt.Errorf("edge %d: source task '%s' not found", edgeIdx, edge.From.Task)
		}
		toTask, ok := ps.Tasks[edge.To.Task]
		if !ok {
			return fmt.Errorf("edge %d: target task '%s' not found", edgeIdx, edge.To.Task)
		}

		// Check ports exist
		if _, ok := fromTask.Outputs[edge.From.Port]; !ok {
			return fmt.Errorf("edge %d: output port '%s' not found on task '%s'", edgeIdx, edge.From.Port, edge.From.Task)
		}
		if _, ok := toTask.Inputs[edge.To.Port]; !ok {
			return fmt.Errorf("edge %d: input port '%s' not found on task '%s'", edgeIdx, edge.To.Port, edge.To.Task)
		}
	}

	return nil
}

// ToJSON serializes the PipelineSpec to JSON.
func (ps *PipelineSpec) ToJSON() ([]byte, error) {
	return json.MarshalIndent(ps, "", "  ")
}

// FromJSON deserializes a PipelineSpec from JSON.
func FromJSON(data []byte) (*PipelineSpec, error) {
	var spec PipelineSpec
	if err := json.Unmarshal(data, &spec); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
	}
	return &spec, nil
}

File: sdk/flowforge/ir/spec.py
"""FlowForge IR Python module - dataclasses and builders for pipeline specifications."""

from __future__ import annotations
from dataclasses import dataclass, field, asdict
from typing import Dict, List, Optional, Any
from enum import Enum
import json


class TaskType(str, Enum):
    """Task type enumeration."""
    SOURCE = "Source"
    TRANSFORM = "Transform"
    SINK = "Sink"
    CONDITIONAL = "Conditional"
    RETRY = "Retry"
    SCHEDULE = "Schedule"


@dataclass
class PipelineMetadata:
    """Pipeline metadata."""
    name: str
    version: Optional[str] = None
    namespace: Optional[str] = None
    owner: Optional[str] = None
    description: Optional[str] = None
    tags: Dict[str, str] = field(default_factory=dict)


@dataclass
class Handler:
    """Handler specification."""
    type: str  # python, sql, spark, docker, http
    source: str
    env: Dict[str, str] = field(default_factory=dict)


@dataclass
class RetryPolicy:
    """Retry configuration."""
    max_attempts: int
    backoff: str = "fixed"  # linear, exponential, fixed
    backoff_multiplier: float = 1.0
    initial_delay_seconds: int = 1


@dataclass
class CostDimension:
    """Single cost dimension."""
    unit: str
    quantity: float


@dataclass
class CostEstimate:
    """Cost estimation."""
    compute: Optional[CostDimension] = None
    storage: Optional[CostDimension] = None
    network: Optional[CostDimension] = None


@dataclass
class Task:
    """Task specification."""
    type: TaskType
    handler: Handler
    description: Optional[str] = None
    inputs: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    outputs: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    executor_config: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    cost_estimate: Optional[CostEstimate] = None
    retry: Optional[RetryPolicy] = None
    timeout: Optional[str] = None
    metadata: Dict[str, str] = field(default_factory=dict)


@dataclass
class TaskPort:
    """Task port reference."""
    task: str
    port: str


@dataclass
class Edge:
    """Data flow edge."""
    from_task: TaskPort
    to_task: TaskPort


@dataclass
class PipelineSpec:
    """Complete pipeline specification."""
    api_version: str = "flowforge.io/v1"
    kind: str = "Pipeline"
    metadata: Optional[PipelineMetadata] = None
    tasks: Dict[str, Task] = field(default_factory=dict)
    edges: List[Edge] = field(default_factory=list)

    def validate(self) -> None:
        """Validate the pipeline specification."""
        if self.api_version != "flowforge.io/v1":
            raise ValueError(f"Invalid apiVersion: {self.api_version}")
        if self.kind != "Pipeline":
            raise ValueError(f"Invalid kind: {self.kind}")
        if not self.metadata or not self.metadata.name:
            raise ValueError("metadata.name is required")
        if not self.tasks:
            raise ValueError("At least one task is required")

        # Validate edges reference valid tasks and ports
        for i, edge in enumerate(self.edges):
            if edge.from_task.task not in self.tasks:
                raise ValueError(f"Edge {i}: source task '{edge.from_task.task}' not found")
            if edge.to_task.task not in self.tasks:
                raise ValueError(f"Edge {i}: target task '{edge.to_task.task}' not found")

            from_task = self.tasks[edge.from_task.task]
            to_task = self.tasks[edge.to_task.task]

            if edge.from_task.port not in from_task.outputs:
                raise ValueError(
                    f"Edge {i}: output port '{edge.from_task.port}' "
                    f"not found on task '{edge.from_task.task}'"
                )
            if edge.to_task.port not in to_task.inputs:
                raise ValueError(
                    f"Edge {i}: input port '{edge.to_task.port}' "
                    f"not found on task '{edge.to_task.task}'"
                )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "apiVersion": self.api_version,
            "kind": self.kind,
            "metadata": asdict(self.metadata) if self.metadata else {},
            "tasks": {
                task_id: self._task_to_dict(task)
                for task_id, task in self.tasks.items()
            },
            "edges": [
                {
                    "from": {"task": edge.from_task.task, "port": edge.from_task.port},
                    "to": {"task": edge.to_task.task, "port": edge.to_task.port},
                }
                for edge in self.edges
            ],
        }

    def to_json(self) -> str:
        """Serialize to JSON."""
        return json.dumps(self.to_dict(), indent=2)

    @staticmethod
    def _task_to_dict(task: Task) -> Dict[str, Any]:
        """Convert task to dictionary."""
        result = {
            "type": task.type.value,
            "handler": {"type": task.handler.type, "source": task.handler.source},
        }
        if task.handler.env:
            result["handler"]["env"] = task.handler.env
        if task.description:
            result["description"] = task.description
        if task.inputs:
            result["inputs"] = task.inputs
        if task.outputs:
            result["outputs"] = task.outputs
        if task.executor_config:
            result["executorConfig"] = task.executor_config
        if task.cost_estimate:
            result["costEstimate"] = asdict(task.cost_estimate)
        if task.retry:
            result["retry"] = {
                "maxAttempts": task.retry.max_attempts,
                "backoff": task.retry.backoff,
                "backoffMultiplier": task.retry.backoff_multiplier,
                "initialDelaySeconds": task.retry.initial_delay_seconds,
            }
        if task.timeout:
            result["timeout"] = task.timeout
        if task.metadata:
            result["metadata"] = task.metadata
        return result

    @staticmethod
    def from_json(json_str: str) -> PipelineSpec:
        """Deserialize from JSON."""
        data = json.loads(json_str)
        return PipelineSpec.from_dict(data)

    @staticmethod
    def from_dict(data: Dict[str, Any]) -> PipelineSpec:
        """Create from dictionary."""
        metadata_data = data.get("metadata", {})
        metadata = PipelineMetadata(
            name=metadata_data.get("name"),
            version=metadata_data.get("version"),
            namespace=metadata_data.get("namespace"),
            owner=metadata_data.get("owner"),
            description=metadata_data.get("description"),
            tags=metadata_data.get("tags", {}),
        )

        tasks = {}
        for task_id, task_data in data.get("tasks", {}).items():
            handler_data = task_data.get("handler", {})
            handler = Handler(
                type=handler_data.get("type"),
                source=handler_data.get("source"),
                env=handler_data.get("env", {}),
            )

            retry_data = task_data.get("retry")
            retry = None
            if retry_data:
                retry = RetryPolicy(
                    max_attempts=retry_data.get("maxAttempts", 1),
                    backoff=retry_data.get("backoff", "fixed"),
                    backoff_multiplier=retry_data.get("backoffMultiplier", 1.0),
                    initial_delay_seconds=retry_data.get("initialDelaySeconds", 1),
                )

            cost_data = task_data.get("costEstimate")
            cost_estimate = None
            if cost_data:
                cost_estimate = CostEstimate(
                    compute=CostDimension(**cost_data["compute"]) if cost_data.get("compute") else None,
                    storage=CostDimension(**cost_data["storage"]) if cost_data.get("storage") else None,
                    network=CostDimension(**cost_data["network"]) if cost_data.get("network") else None,
                )

            task = Task(
                type=TaskType(task_data.get("type")),
                handler=handler,
                description=task_data.get("description"),
                inputs=task_data.get("inputs", {}),
                outputs=task_data.get("outputs", {}),
                executor_config=task_data.get("executorConfig", {}),
                cost_estimate=cost_estimate,
                retry=retry,
                timeout=task_data.get("timeout"),
                metadata=task_data.get("metadata", {}),
            )
            tasks[task_id] = task

        edges = []
        for edge_data in data.get("edges", []):
            from_data = edge_data.get("from", {})
            to_data = edge_data.get("to", {})
            edge = Edge(
                from_task=TaskPort(task=from_data.get("task"), port=from_data.get("port")),
                to_task=TaskPort(task=to_data.get("task"), port=to_data.get("port")),
            )
            edges.append(edge)

        spec = PipelineSpec(
            api_version=data.get("apiVersion", "flowforge.io/v1"),
            kind=data.get("kind", "Pipeline"),
            metadata=metadata,
            tasks=tasks,
            edges=edges,
        )
        return spec
